File: src/gameStates/StateHandler.java
package gameStates;

import main.Main;

import java.awt.*;
import java.awt.event.KeyEvent;

public class StateHandler {
    private GameState currState;
    private Main main;
    public volatile boolean running = true;
    public Object lock;
    private GameStateFactory stateFactory;

    public StateHandler(Main main) {
        this.main = main;
        stateFactory = new GameStateFactory(this);
        setState(GameStateID.Level1);
        lock = new Object();
    }

    public void setState(GameStateID stateId) {
        currState = stateFactory.getState(stateId);
    }

    public synchronized void tick() {
        currState.tick();
    }

    public synchronized void render(Graphics g) {
        currState.render(g);
    }

    public synchronized void keyPressed(int k) throws InterruptedException {
        //General events
        if (k == KeyEvent.VK_ESCAPE) {
            main.stopProgram();
        }
        currState.keyPressed(k);
    }

    public void keyReleased(int k) {
        currState.keyReleased(k);
    }
}

File: public/game/StateHandler.js
'use strict'
class StateHandler {
    #currState

    constructor() {
        this.setState('LEVEL1')
    }

    tick(delta) {
        this.#currState.tick(delta)
    }

    render(ctx) {
        this.#currState.render(ctx)
    }

    setState(state) {
        if (state === 'MAINMENU') {
            this.#currState = new MainMenu(this)
        } else if (state === 'INTRO') {
            this.#currState = new Intro(this)
        } else if (state === 'LEVEL1') {
            this.#currState = new Level1(this)
        } else if (state === 'GAMEOVER') {
            this.#currState = new GameOver(this)
        } else if (state === 'WIN') {
            this.#currState = new Win(this)
        } else {
            console.error("GameState '" + state + "' doesn't exist")
        }
    }

    keyPressed(keyCode) {
        this.#currState.keyPressed(keyCode)
    }

    keyReleased(keyCode) {
        this.#currState.keyReleased(keyCode)
    }
}
